File: hackerrank/problem solving/algorithms/2 - implementation/34 - cut the sticks/oo/cut_the_sticks_oo.rb
# Source: https://www.hackerrank.com/challenges/cut-the-sticks/problem?isFullScreen=true

def main
    _size = readANumber
    array = readAnArray
    array = array.sort

    obj = CutTheSticks.new array
    result = obj.getRemainingSticksLengthArray
    puts result
end

    def readANumber
        gets.to_i
    end

    def readAnArray
        gets.split.map(&:to_i)
    end

    class CutTheSticks
        @sticksLengthArray
        @sizeSticksLengthArray
        @remainingSticksLengthArray

        def initialize array
            @sticksLengthArray = array
            @sizeSticksLengthArray = array.size
            _initializeRemainingSticksLengthArray
            _cutTheSticks
        end

            private def _initializeRemainingSticksLengthArray
                @remainingSticksLengthArray = [@sizeSticksLengthArray]
            end

            private def _cutTheSticks
                i = 0
                while i < @sizeSticksLengthArray
                    shortestStickLength = @sticksLengthArray[i]

                    i = _findFirstElementIndexDifferentOfFirstCurrentShortestStickLengthIndex i
                    currentIterationSize = @sizeSticksLengthArray - i
                    if i == @sizeSticksLengthArray
                        break
                    end

                    @remainingSticksLengthArray.append currentIterationSize
                    _decreaseSticksLengthsHigherThanCurrentShortestStickLength i, shortestStickLength
                end
            end

                private def _findFirstElementIndexDifferentOfFirstCurrentShortestStickLengthIndex index
                    shortestStickLength = @sticksLengthArray[index]

                    while index < @sizeSticksLengthArray and @sticksLengthArray[index] == shortestStickLength
                        index += 1
                    end

                    index
                end

                private def _decreaseSticksLengthsHigherThanCurrentShortestStickLength index, shortestStickLength
                    for i in index..(@sizeSticksLengthArray-1)
                        @sticksLengthArray[i] -= shortestStickLength
                    end
                end

        public def getRemainingSticksLengthArray
            @remainingSticksLengthArray
        end
    end

main

File: hackerrank/problem solving/algorithms/2 - implementation/34 - cut the sticks/oo/cut_the_sticks_oo.cs
// Source: https://www.hackerrank.com/challenges/cut-the-sticks/problem?isFullScreen=true

using System;

public class Solution
{
    public static void Main()
    {
        int _size = _ReadANumber();
        List<int> array = _ReadAnArray();
        array.Sort();

        CutTheSticks obj = new CutTheSticks(array);
        List<int> result = obj.GetRemainingSticksLengthArray();
        printArray(result);
    }

        private static int _ReadANumber()
        {
            int number = int.Parse(Console.ReadLine());
            return number;
        }

        private static List<int> _ReadAnArray()
        {
            List<int> array = Console.ReadLine().Split().Select(int.Parse).ToList();
            return array;
        }

        public static void printArray(List<int> array)
        {
            array.ForEach(Console.WriteLine);
        }

}

    public class CutTheSticks
    {
        private List<int> _SticksLengthArray;
        private int _SizeSticksLengthArray;
        private List<int> _RemainingSticksLengthArray;

        public CutTheSticks(List<int> array)
        {
            _SticksLengthArray = array;
            _SizeSticksLengthArray = array.Count;
            _InitializeRemainingSticksLengthArray();
            _CutTheSticks();
        }

            private void _InitializeRemainingSticksLengthArray()
            {
                _RemainingSticksLengthArray = new List<int> {_SizeSticksLengthArray};
            }

            private void _CutTheSticks()
            {
                for (int i = 0; i < _SizeSticksLengthArray; )
                {
                    int shortestStickLength = _SticksLengthArray[i];

                    i = _FindFirstElementIndexDifferentOfFirstCurrentShortestStickLengthIndex(i);
                    int currentIterationSize = _SizeSticksLengthArray - i;
                    if (i == _SizeSticksLengthArray)
                        break;

                    _RemainingSticksLengthArray.Add(currentIterationSize);
                    _DecreaseSticksLengthsHigherThanCurrentShortestStickLength(i, shortestStickLength);
                }
            }

                private int _FindFirstElementIndexDifferentOfFirstCurrentShortestStickLengthIndex(int index)
                {
                    int shortestStickLength = _SticksLengthArray[index];
                    while (index < _SizeSticksLengthArray && _SticksLengthArray[index] == shortestStickLength)
                        index++;
                    return index;
                }

                private void _DecreaseSticksLengthsHigherThanCurrentShortestStickLength(int index, int shortestStickLength)
                {
                    _SticksLengthArray = _SticksLengthArray.Select(element => element - shortestStickLength).ToList();
                }

        public List<int> GetRemainingSticksLengthArray()
        {
            return _RemainingSticksLengthArray;
        }
    }
